### PR TITLE
build: produce race-enabled test binaries

### PR DESCRIPTION
**Pull request by tamird**
Updates #8897.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9742)

### CODE REVIEW

File: build/build-static-binaries.sh [hunk 1]
```diff
@@ -9,7 +9,13 @@ tags=${2-}
 source "$(dirname "${0}")"/build-common.sh
 
 time make STATIC=1 build
+# Build test binaries. Note that the acceptance binary will not be included in
+# the archive, but is instead uploaded directly by push-aws.sh.
+#
+# TODO(tamird,jordan): consider running race-enabled acceptance tests to
+# expose data races in the acceptance tests themselves.
 time make STATIC=1 testbuild PKG=./... TAGS="$tags acceptance"
+time make STATIC=1 testbuild PKG=./... TAGS="$tags acceptance" GOFLAGS=-race TESTBUILDPREFIX=race.
```
Comment (tschottdorf) [inline build/build-static-binaries.sh]: doubles the build time, no? I'm worried that once I add `deadlock` in here we're going to wait **forever** for the `master` binaries to bake. How do you feel about doing this in a nightly instead (with the baking succeeding a requirement for running the nightlies that need these binaries).

Comment (tamird) [inline build/build-static-binaries.sh]: Agreed. Moved to build-race-binaries in a new commit; PTAL and I'll squash if it looks good to you.

**Comment by tschottdorf** [comment]
Looks good.

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[Makefile, line 84 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCB-MoZOb8rp5cs2y:-KTKCB-MoZOb8rp5cs2z:b-b411ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/9535ba7126737bf9e9f9c60f9d54f87f8f0da223/Makefile#L84)):*
> ```Makefile
> 	@$(GO) $(BUILDMODE) -v $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)'
> 
> TESTBUILDPREFIX :=
> ```

Add a comment.

---

*[build/build-race-binaries.sh, line 19 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCb69JmZAO0_DFIHf:-KTKCb69JmZAO0_DFIHg:ba7is03) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be9b5b43aa4e9e661e46771ef244aeb7609342e/build/build-race-binaries.sh#L19)):*
> ```Shell
> 
> # We don't check all test binaries, but one from each invocation.
> check_static cockroach
> ```

Where does `cockroach` come from? And if so, wouldn't we want to build `race.cockroach` (which should be called `cockroach.race` unless we want to change the convention)?

---

*[build/build-static-binaries.sh, line 17 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCK6VlY1c1Hn_lU_R:-KTKCK6VlY1c1Hn_lU_S:b-ojqbek) ([raw file](https://github.com/cockroachdb/cockroach/blob/9535ba7126737bf9e9f9c60f9d54f87f8f0da223/build/build-static-binaries.sh#L17)):*
> ```Shell
> # expose data races in the acceptance tests themselves.
> time make STATIC=1 testbuild PKG=./... TAGS=acceptance
> time make STATIC=1 testbuild PKG=./... TAGS=acceptance GOFLAGS=-race TESTBUILDPREFIX=race.
> ```

There's precedent for making this a suffix `.race` instead of a prefix (`rho` cluster).

---

*[build/build-static-binaries.sh, line 7 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKBjax6N3cqHVHwWZb:-KTKBjay_2viL3tSphbZ:b-5nezhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432e7be3c7efb4601a0d33f961ca392ad0805a7/build/build-static-binaries.sh#L7)):*
> ```Shell
> 
> archive=$1
> tags=${2-}
> ```

Are you sure no caller relies on this (I never saw it supplied, but stilL) ? Should make it illegal to supply `${2-}` with a comment stipulating the removal if it hasn't failed by `now()+1month`.

---

*[build/build-static-binaries.sh, line 12 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKC0bEkxxSB_sslRVc:-KTKC0bEkxxSB_sslRVd:b-u8pirv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be9b5b43aa4e9e661e46771ef244aeb7609342e/build/build-static-binaries.sh#L12)):*
> ```Shell
> time make STATIC=1 build
> # Build test binaries. Note that the acceptance binary will not be included in
> # the archive, but is instead uploaded directly by push-aws.sh.
> ```

No good reason for that, right? (Not for this PR).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-:-KTKCwFgRsgR_FS4Rykc:b-vs1pr4)*

**Comment by tamird** [comment]
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[Makefile, line 84 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCB-MoZOb8rp5cs2y:-KTKEsw5p2a_leW1_BFD:b-f0d78o) ([raw file](https://github.com/cockroachdb/cockroach/blob/9535ba7126737bf9e9f9c60f9d54f87f8f0da223/Makefile#L84)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Add a comment.
</details>
Renamed to "suffix" which is hopefully more intuitive.

---

*[build/build-race-binaries.sh, line 19 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCb69JmZAO0_DFIHf:-KTKFIbU1g9sERRZBQvE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be9b5b43aa4e9e661e46771ef244aeb7609342e/build/build-race-binaries.sh#L19)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Where does `cockroach` come from? And if so, wouldn't we want to build `race.cockroach` (which should be called `cockroach.race` unless we want to change the convention)?
</details>
Done.

---

*[build/build-static-binaries.sh, line 17 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCK6VlY1c1Hn_lU_R:-KTKFA6-4CFQWdViqKRO:b9vo5lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9535ba7126737bf9e9f9c60f9d54f87f8f0da223/build/build-static-binaries.sh#L17)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> There's precedent for making this a suffix `.race` instead of a prefix (`rho` cluster).
</details>
Made it a suffix, though it's likely to break whatever script is build `rho` cluster binaries in TC (since the suffix will now come from build-race-binaries). cc @jordanlewis 

---

*[build/build-static-binaries.sh, line 7 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKBjax6N3cqHVHwWZb:-KTKEvnhm5J7P9OQrOFx:b-j8e0xp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432e7be3c7efb4601a0d33f961ca392ad0805a7/build/build-static-binaries.sh#L7)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Are you sure no caller relies on this (I never saw it supplied, but stilL) ? Should make it illegal to supply `${2-}` with a comment stipulating the removal if it hasn't failed by `now()+1month`.
</details>
See the commit message. It was used to build stdmalloc, which I think we no longer do. @jordanlewis is the authority.

---

*[build/build-static-binaries.sh, line 12 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKC0bEkxxSB_sslRVc:-KTKF1ZYCxJFV55BvAY0:b-ovgm6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6be9b5b43aa4e9e661e46771ef244aeb7609342e/build/build-static-binaries.sh#L12)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> No good reason for that, right? (Not for this PR).
</details>
Yes and no. The reason this is done is so that the stress runner can blindly run everything that looks like a test, and we don't want it running acceptance tests. We could do the filtering runner-side, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742)*

**Comment by tschottdorf** [comment]
:lgtm:, @jordanlewis should also take a look (and perhaps help with the transition).

BTW, if you don't mind adding a corresponding `deadlock` facility (tests and binary), that'd be great, otherwise I'll do it at  some point. I have a feeling that a deadlock-enabled cluster will also show some lock order violations as our test coverage for those isn't sufficient.

---

Reviewed 2 of 5 files at r5, 8 of 8 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-:-KTKGR5szfY_y5Ac4hvm:b-ulzuzs)*

**Comment by tamird** [comment]
Yeah, I'll add that in a follow-up, after I coordinate with @jordanlewis to
remove build-race-binaries.

On Wed, Oct 5, 2016 at 10:37 AM, Tobias Schottdorf <notifications@github.com

**Comment by tschottdorf** [comment]
Great, thanks.

**Comment by jordanlewis** [comment]
:lgtm: apart from small naming nit

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[Makefile, line 96 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKHpe2R0WgGGvCORMW:-KTKHpe39sUkcPRylXmf:b-8r5akv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7c1c3f10781959762415e1deea51056264726c0/Makefile#L96)):*
> ```Makefile
> 	$(SHELL)
> 
> .PHONY: ideps
> ```

`ideps`? could you make this more descriptive? i gather it means "install deps" but it's not that clear - maybe `testdeps` captures the meaning better?

---

*[build/build-static-binaries.sh, line 17 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCK6VlY1c1Hn_lU_R:-KTKItWnVf0xds4KwReI:bpykydi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9535ba7126737bf9e9f9c60f9d54f87f8f0da223/build/build-static-binaries.sh#L17)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Made it a suffix, though it's likely to break whatever script is build `rho` cluster binaries in TC (since the suffix will now come from build-race-binaries). cc @jordanlewis 
</details>
This comment is out of date I think.

---

*[build/build-static-binaries.sh, line 7 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKBjax6N3cqHVHwWZb:-KTKIY6jrFfLrvaA1UiH:b-9gvxtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1432e7be3c7efb4601a0d33f961ca392ad0805a7/build/build-static-binaries.sh#L7)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> See the commit message. It was used to build stdmalloc, which I think we no longer do. @jordanlewis is the authority.
</details>
It's true `stdmalloc` is no longer used or supported. I don't think TeamCity uses this argument. If there are other callers they may break but I think that's okay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-:-KTKJAFrYxMy1f_pdnt7:b-r6agbv)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[Makefile, line 96 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKHpe2R0WgGGvCORMW:-KTKJiBkwDBf7WLapVrf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7c1c3f10781959762415e1deea51056264726c0/Makefile#L96)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> `ideps`? could you make this more descriptive? i gather it means "install deps" but it's not that clear - maybe `testdeps` captures the meaning better?
</details>
Done.

---

*[build/build-static-binaries.sh, line 17 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCK6VlY1c1Hn_lU_R:-KTKJXQxK8FS14S9kRf8:b-xjv0hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9535ba7126737bf9e9f9c60f9d54f87f8f0da223/build/build-static-binaries.sh#L17)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> This comment is out of date I think.
</details>
Which comment? The acceptance test? No, it's correct; this has always been the behaviour.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742)*

**Comment by jordanlewis** [comment]
Review status: 2 of 10 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[build/build-static-binaries.sh, line 17 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-KTKCK6VlY1c1Hn_lU_R:-KTKKb9xZIGhpn2UNDy2:b-58pe91) ([raw file](https://github.com/cockroachdb/cockroach/blob/9535ba7126737bf9e9f9c60f9d54f87f8f0da223/build/build-static-binaries.sh#L17)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Which comment? The acceptance test? No, it's correct; this has always been the behaviour.
</details>
Sorry I meant the reviewable comments. Code comment LGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742)*

**Comment by jordanlewis** [comment]
Reviewed 2 of 5 files at r5, 8 of 8 files at r11.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742)*

**Comment by tamird** [comment]
OK, updated the TC config to look for cockroach.race instead of `cockroach`. will merge on green.

**Comment by tamird** [comment]
Pushed another update that removes build-race-binaries. I've made the necessary update to TC.

@tschottdorf PTAL since you can build your deadlock binaries with a simple:

```
build/builder.sh TAGS=deadlock SUFFIX=.deadlock build/build-static-binaries.sh
```

**Comment by tschottdorf** [comment]
:lgtm:

---

Reviewed 8 of 8 files at r15.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9742#-:-KTKsOZQu-2D5qj2gWqx:bnfp4nl)*

**Pull request merged**